src/java/org/apache/cassandra/index/sai/disk/format/IndexDescriptor.java

```diff
@@ -0,0 +1,530 @@
+    }
+
+    public String componentName(IndexComponent indexComponent, IndexContext indexContext)
+    {
+        return version.fileNameFormatter().format(indexComponent, indexContext);
+    }
+
+    public File fileFor(IndexComponent component)
+    {
+        return onDiskPerSSTableFileMap.computeIfAbsent(component, c -> createFile(c, null, false));
+    }
+
+    public File fileFor(IndexComponent component, IndexContext indexContext)
+    {
+        return onDiskPerIndexFileMap.computeIfAbsent(Pair.create(component, indexContext.getIndexName()),
+                                                     p -> createFile(component, indexContext, false));
+    }
+
+    public Set<Component> getLivePerSSTableComponents()
+    {
+        registerPerSSTableComponents();
+        return perSSTableComponents.stream()
+                                   .map(c -> new Component(Component.Type.CUSTOM, componentName(c)))
+                                   .collect(Collectors.toSet());
+    }
+
+    public Set<Component> getLivePerIndexComponents(IndexContext indexContext)
+    {
+        registerPerIndexComponents(indexContext);
+        return perIndexComponents.containsKey(indexContext.getIndexName())
+               ? perIndexComponents.get(indexContext.getIndexName())
+                                   .stream()
+                                   .map(c -> new Component(Component.Type.CUSTOM, componentName(c, indexContext)))
+                                                                         .collect(Collectors.toSet())
+                                                     : Collections.emptySet();
+    }
+
+    public PrimaryKeyMap.Factory newPrimaryKeyMapFactory(SSTableReader sstable)
+    {
+        return version.onDiskFormat().newPrimaryKeyMapFactory(this, sstable);
+    }
+
+    public SearchableIndex newSearchableIndex(SSTableContext sstableContext, IndexContext indexContext)
+    {
+        return version.onDiskFormat().newSearchableIndex(sstableContext, indexContext);
+    }
+
+    public PerSSTableWriter newPerSSTableWriter() throws IOException
+    {
+        return version.onDiskFormat().newPerSSTableWriter(this);
+    }
+
+    public PerIndexWriter newPerIndexWriter(StorageAttachedIndex index,
+                                            LifecycleNewTracker tracker,
+                                            RowMapping rowMapping)
+    {
+        return version.onDiskFormat().newPerIndexWriter(index, this, tracker, rowMapping);
+    }
+
+    public boolean isPerSSTableBuildComplete()
+    {
+        return version.onDiskFormat().isPerSSTableBuildComplete(this);
+    }
+
+    public boolean isPerIndexBuildComplete(IndexContext indexContext)
+    {
+        return version.onDiskFormat().isPerIndexBuildComplete(this, indexContext);
+    }
+
+    public boolean isIndexEmpty(IndexContext indexContext)
+    {
+        return isPerIndexBuildComplete(indexContext) && numberOfComponents(indexContext) == 1;
```

Comment:
nit: maybe a quick comment on why this is 1

Reply:
Note: I have renamed `numberOfComponents` to `numberOfPerIndexComponents` to make it clearer.